# ABOUT DATASET
This dataset shows the Battery Electric Vehicles (BEVs) and Plug-in Hybrid Electric Vehicles (PHEVs) that are currently registered through Washington State Department of Licensing (DOL) updated 2024.

# BUSINESS PROBLEM

As electric vehicle (EV) adoption increases in Washington State, policymakers, utility providers, and automakers need data-driven insights to support infrastructure planning, incentive programs, and customer outreach. However, understanding where and how EVs are being adopted—by city, model, year, and utility provider—is fragmented.

## Key Problem:
How can stakeholders identify trends, disparities, and opportunities in EV adoption across geographic and demographic regions in Washington State to optimize resource allocation and policy development?

# Business Objectives

### Adoption Trend Analysis: 
Identify trends in EV adoption over time, by make, model, and model year.

### Geographic Distribution:
Analyze EV ownership density by county, city, and legislative district.

### Infrastructure Planning:
Identify areas with high EV population but limited utility coverage or support for EV infrastructure.

### Incentive Effectiveness:
Evaluate the correlation between CAFV eligibility and EV adoption.

### Equity Assessment:
Compare EV adoption across different 2020 census tracts to uncover demographic or economic disparities.

# Proposed Methodology
1. ### Data Cleaning and Preprocessing
Handle missing values (e.g., Electric Range = 0)

Parse and standardize geographical coordinates and utilities

Convert categorical fields (e.g., Make, Model, CAFV Eligibility) to appropriate formats

2. ### Exploratory Data Analysis (EDA)
Temporal trends (Model Year vs Count)

Geospatial mapping of EV ownership (using coordinates, county, city)

Popular makes/models and their distribution

EV range and MSRP trends over years

3. ### Geospatial Analysis
Map EV density using Vehicle Location

Overlay electric utility regions to assess alignment with EV adoption

Identify underserved areas (e.g., low EV counts with high population)

4. ### CAFV and Policy Impact Evaluation
Cross-tabulation of CAFV eligibility vs. adoption by year

Visualize the impact of eligibility status on EV model distribution

5. ### Equity & Access Analysis
Join dataset with census data using 2020 Census Tract

Examine demographic characteristics (income, race, education) and EV adoption levels

6. ### Recommendations
Highlight underserved regions for infrastructure build-out

Recommend adjustments to incentive programs based on adoption and eligibility data

Inform utility providers on capacity planning and future demand areas



# DATA UNDERSTANDING

## Load and Inspect the data

In [1]:
# Import necessary Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [2]:
# Loading the dataset
df = pd.read_csv("Electric_Vehicle_Population_Data.csv")

In [3]:
# Check the data head
df.head()

VIN (1-10)     County       City State  Postal Code  Model Year   Make  \
0  5YJYGDEE1L       King    Seattle    WA      98122.0        2020  TESLA   
1  7SAYGDEE9P  Snohomish    Bothell    WA      98021.0        2023  TESLA   
2  5YJSA1E4XK       King    Seattle    WA      98109.0        2019  TESLA   
3  5YJSA1E27G       King   Issaquah    WA      98027.0        2016  TESLA   
4  5YJYGDEE5M     Kitsap  Suquamish    WA      98392.0        2021  TESLA   

     Model           Electric Vehicle Type  \
0  MODEL Y  Battery Electric Vehicle (BEV)   
1  MODEL Y  Battery Electric Vehicle (BEV)   
2  MODEL S  Battery Electric Vehicle (BEV)   
3  MODEL S  Battery Electric Vehicle (BEV)   
4  MODEL Y  Battery Electric Vehicle (BEV)   

   Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0            Clean Alternative Fuel Vehicle Eligible             291   
1  Eligibility unknown as battery range has not b...               0   
2            Clean Alternative Fuel Vehicle Eligible             270   
3            Clean Alternative Fuel Vehicle Eligible             210   
4  Eligibility unknown as battery range has not b...               0   

   Base MSRP  Legislative District  DOL Vehicle ID  \
0          0                  37.0       125701579   
1          0                   1.0       244285107   
2          0                  36.0       156773144   
3          0                   5.0       165103011   
4          0                  23.0       205138552   

                Vehicle Location  \
0   POINT (-122.30839 47.610365)   
1  POINT (-122.179458 47.802589)   
2   POINT (-122.34848 47.632405)   
3   POINT (-122.03646 47.534065)   
4   POINT (-122.55717 47.733415)   

                                Electric Utility  2020 Census Tract  
0   CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)       5.303301e+10  
1                         PUGET SOUND ENERGY INC       5.306105e+10  
2   CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)       5.303301e+10  
3  PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303303e+10  
4                         PUGET SOUND ENERGY INC       5.303594e+10

Based on the above code output it shows 5 Tesla electric vehicles registered in Washington State, primarily in King County. All are Battery Electric Vehicles (BEVs), with models including the Model Y and Model S.

However, there are inconsistencies in the data:

Electric Range of 0 appears in 3 out of 5 entries, which is likely a garbage or missing value, especially for models like the Model Y or S, which typically have substantial range.

Base MSRP is 0 for all entries, suggesting incomplete or unreported pricing data.

The "CAFV Eligibility" is listed as "unknown" for some vehicles, indicating a lack of battery range data.

This suggests a need for data cleaning, particularly for range and price fields, before any meaningful analysis or policy decisions can be made from this dataset.

In [4]:
# Check the data head
df.tail()

VIN (1-10)  County           City State  Postal Code  Model Year  \
177861  7SAYGDEE3N  Pierce    Bonney Lake    WA      98391.0        2022   
177862  KM8K23AG1P   Mason        Shelton    WA      98584.0        2023   
177863  5YJYGDEE6M   Grant         Quincy    WA      98848.0        2021   
177864  WVGKMPE27M    King  Black Diamond    WA      98010.0        2021   
177865  5YJ3E1EA8M  Pierce         Tacoma    WA      98422.0        2021   

              Make          Model           Electric Vehicle Type  \
177861       TESLA        MODEL Y  Battery Electric Vehicle (BEV)   
177862     HYUNDAI  KONA ELECTRIC  Battery Electric Vehicle (BEV)   
177863       TESLA        MODEL Y  Battery Electric Vehicle (BEV)   
177864  VOLKSWAGEN           ID.4  Battery Electric Vehicle (BEV)   
177865       TESLA        MODEL 3  Battery Electric Vehicle (BEV)   

        Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
177861  Eligibility unknown as battery range has not b...               0   
177862  Eligibility unknown as battery range has not b...               0   
177863  Eligibility unknown as battery range has not b...               0   
177864  Eligibility unknown as battery range has not b...               0   
177865  Eligibility unknown as battery range has not b...               0   

        Base MSRP  Legislative District  DOL Vehicle ID  \
177861          0                  31.0       195224452   
177862          0                  35.0       228454180   
177863          0                  13.0       168797219   
177864          0                   5.0       182448801   
177865          0                  27.0       211464683   

                       Vehicle Location  \
177861     POINT (-122.183805 47.18062)   
177862    POINT (-123.105305 47.211085)   
177863  POINT (-119.8493873 47.2339933)   
177864     POINT (-122.00451 47.312185)   
177865      POINT (-122.38578 47.28971)   

                                         Electric Utility  2020 Census Tract  
177861      PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.305307e+10  
177862  BONNEVILLE POWER ADMINISTRATION||CITY OF TACOM...       5.304596e+10  
177863                           PUD NO 2 OF GRANT COUNTY       5.302501e+10  
177864      PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303303e+10  
177865  BONNEVILLE POWER ADMINISTRATION||CITY OF TACOM...       5.305394e+10

All five electric vehicles listed are recent Battery Electric Vehicles (BEVs) (model years 2021–2023) from brands like Tesla, Hyundai, and Volkswagen, registered across various Washington counties.

However, all entries have missing or zero values for:

Electric Range

Base MSRP

CAFV Eligibility

This strongly indicates incomplete or faulty data entry, making the records unreliable for any analysis involving vehicle performance, eligibility, or pricing. Data validation and enrichment are needed before further use.

In [5]:
# Check the data shape
df.shape

(177866, 17)

The output (177866, 17) indicates the shape of the dataset — specifically:

177,866 rows (individual electric vehicle records)

17 columns (features/attributes per vehicle)

🔍 Quick Insight:
The dataset is large and comprehensive, suggesting it's a valuable resource for analyzing electric vehicle trends in Washington State. However, based on the earlier sample rows, data quality issues like missing values e.g. in electric range, MSRP etc may need to be addressed before conducting accurate analyses.

In [6]:
# Chack the data size
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 177866 entries, 0 to 177865
Data columns (total 17 columns):
 #   Column                                             Non-Null Count   Dtype  
---  ------                                             --------------   -----  
 0   VIN (1-10)                                         177866 non-null  object 
 1   County                                             177861 non-null  object 
 2   City                                               177861 non-null  object 
 3   State                                              177866 non-null  object 
 4   Postal Code                                        177861 non-null  float64
 5   Model Year                                         177866 non-null  int64  
 6   Make                                               177866 non-null  object 
 7   Model                                              177866 non-null  object 
 8   Electric Vehicle Type                              177866 non-null  object

The dataset contains 177,866 electric vehicle records across 17 columns. While most fields are complete, there are some minor but notable gaps:

Missing values in key location fields like:

County, City, Postal Code (5 missing)

Legislative District (~389 missing)

Vehicle Location, Electric Utility, 2020 Census Tract (5–9 missing)

Data types are appropriate:

Categorical/object: For vehicle details (e.g., Make, Model, CAFV eligibility)

Numerical: For year, range, price, district, etc.

⚠️ Key Flag:
Even though Electric Range and Base MSRP have no nulls, earlier samples show values of 0, indicating garbage or placeholder data that should be cleaned despite appearing complete.

In [7]:
#Discriptive statistics
df.describe()

Postal Code     Model Year  Electric Range      Base MSRP  \
count  177861.000000  177866.000000   177866.000000  177866.000000   
mean    98172.453506    2020.515512       58.842162    1073.109363   
std      2442.450668       2.989384       91.981298    8358.624956   
min      1545.000000    1997.000000        0.000000       0.000000   
25%     98052.000000    2019.000000        0.000000       0.000000   
50%     98122.000000    2022.000000        0.000000       0.000000   
75%     98370.000000    2023.000000       75.000000       0.000000   
max     99577.000000    2024.000000      337.000000  845000.000000   

       Legislative District  DOL Vehicle ID  2020 Census Tract  
count         177477.000000    1.778660e+05       1.778610e+05  
mean              29.127481    2.202313e+08       5.297672e+10  
std               14.892169    7.584987e+07       1.578047e+09  
min                1.000000    4.385000e+03       1.001020e+09  
25%               18.000000    1.814743e+08       5.303301e+10  
50%               33.000000    2.282522e+08       5.303303e+10  
75%               42.000000    2.548445e+08       5.305307e+10  
max               49.000000    4.792548e+08       5.603300e+10

The summary reveals serious data quality issues in two critical fields:

🔋 Electric Range:

Median = 0 and 75% percentile = 75 → At least half the vehicles have no recorded range, likely indicating missing or unreported values.

💲 Base MSRP:

Mean = $1,073, Median = $0, 75% percentile = $0 → Almost 75% of entries have zero MSRP, which is implausible and signals widespread missing data.

⚠️ Key Takeaway:
Despite numeric completeness, over half the records have placeholder values (0) for essential fields like range and price, making them unreliable for analysis without imputation or filtering.

In [8]:
#Check the missing values
df.isnull().sum()

VIN (1-10)                                             0
County                                                 5
City                                                   5
State                                                  0
Postal Code                                            5
Model Year                                             0
Make                                                   0
Model                                                  0
Electric Vehicle Type                                  0
Clean Alternative Fuel Vehicle (CAFV) Eligibility      0
Electric Range                                         0
Base MSRP                                              0
Legislative District                                 389
DOL Vehicle ID                                         0
Vehicle Location                                       9
Electric Utility                                       5
2020 Census Tract                                      5
dtype: int64

The dataset is mostly complete, with a few exceptions.

Legislative District is the most problematic column, missing in a significant number of rows (389). This limits geographic policy analysis at the district level.

Only a few rows (5–9) have missing location-related fields like County, City, Postal Code, and Vehicle Location. These can be dropped or filled using related data.

Electric Range = 0 may not be missing, but rather unknown or non-applicable — requires contextual interpretation.

In [10]:
#Finding the dublicates values
df.duplicated().sum()

np.int64(0)

Based onthe above output there is no duplicate values

In [9]:
# Count the number of unique cities and counties
total_counties = df['County'].nunique()
total_cities = df['City'].nunique()
print("Total Counties:", total_counties)
print("Total Cities:", total_cities)

Total Counties: 196
Total Cities: 723


In [11]:
# Analyze the distribution of electric vehicle types
vehicle_type_counts = df['Electric Vehicle Type'].value_counts()
print("Distribution of Electric Vehicle Types:")
print(vehicle_type_counts)

Distribution of Electric Vehicle Types:
Electric Vehicle Type
Battery Electric Vehicle (BEV)            139210
Plug-in Hybrid Electric Vehicle (PHEV)     38656
Name: count, dtype: int64


In [13]:
# Identify the top electric vehicle makes
top_makes = df['Make'].value_counts().head(10)
df

VIN (1-10)     County           City State  Postal Code  Model Year  \
0       5YJYGDEE1L       King        Seattle    WA      98122.0        2020   
1       7SAYGDEE9P  Snohomish        Bothell    WA      98021.0        2023   
2       5YJSA1E4XK       King        Seattle    WA      98109.0        2019   
3       5YJSA1E27G       King       Issaquah    WA      98027.0        2016   
4       5YJYGDEE5M     Kitsap      Suquamish    WA      98392.0        2021   
...            ...        ...            ...   ...          ...         ...   
177861  7SAYGDEE3N     Pierce    Bonney Lake    WA      98391.0        2022   
177862  KM8K23AG1P      Mason        Shelton    WA      98584.0        2023   
177863  5YJYGDEE6M      Grant         Quincy    WA      98848.0        2021   
177864  WVGKMPE27M       King  Black Diamond    WA      98010.0        2021   
177865  5YJ3E1EA8M     Pierce         Tacoma    WA      98422.0        2021   

              Make          Model           Electric Vehicle Type  \
0            TESLA        MODEL Y  Battery Electric Vehicle (BEV)   
1            TESLA        MODEL Y  Battery Electric Vehicle (BEV)   
2            TESLA        MODEL S  Battery Electric Vehicle (BEV)   
3            TESLA        MODEL S  Battery Electric Vehicle (BEV)   
4            TESLA        MODEL Y  Battery Electric Vehicle (BEV)   
...            ...            ...                             ...   
177861       TESLA        MODEL Y  Battery Electric Vehicle (BEV)   
177862     HYUNDAI  KONA ELECTRIC  Battery Electric Vehicle (BEV)   
177863       TESLA        MODEL Y  Battery Electric Vehicle (BEV)   
177864  VOLKSWAGEN           ID.4  Battery Electric Vehicle (BEV)   
177865       TESLA        MODEL 3  Battery Electric Vehicle (BEV)   

        Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0                 Clean Alternative Fuel Vehicle Eligible             291   
1       Eligibility unknown as battery range has not b...               0   
2                 Clean Alternative Fuel Vehicle Eligible             270   
3                 Clean Alternative Fuel Vehicle Eligible             210   
4       Eligibility unknown as battery range has not b...               0   
...                                                   ...             ...   
177861  Eligibility unknown as battery range has not b...               0   
177862  Eligibility unknown as battery range has not b...               0   
177863  Eligibility unknown as battery range has not b...               0   
177864  Eligibility unknown as battery range has not b...               0   
177865  Eligibility unknown as battery range has not b...               0   

        Base MSRP  Legislative District  DOL Vehicle ID  \
0               0                  37.0       125701579   
1               0                   1.0       244285107   
2               0                  36.0       156773144   
3               0                   5.0       165103011   
4               0                  23.0       205138552   
...           ...                   ...             ...   
177861          0                  31.0       195224452   
177862          0                  35.0       228454180   
177863          0                  13.0       168797219   
177864          0                   5.0       182448801   
177865          0                  27.0       211464683   

                       Vehicle Location  \
0          POINT (-122.30839 47.610365)   
1         POINT (-122.179458 47.802589)   
2          POINT (-122.34848 47.632405)   
3          POINT (-122.03646 47.534065)   
4          POINT (-122.55717 47.733415)   
...                                 ...   
177861     POINT (-122.183805 47.18062)   
177862    POINT (-123.105305 47.211085)   
177863  POINT (-119.8493873 47.2339933)   
177864     POINT (-122.00451 47.312185)   
177865      POINT (-122.38578 47.28971)   

                                         Electric Utility  2020 Census Tract  


In [14]:
# Calculate the top 10 most common makes
top_makes = df['Make'].value_counts().head(10)

print("Top 10 most common makes:")
print(top_makes)

Top 10 most common makes:
Make
TESLA         79659
NISSAN        13998
CHEVROLET     13678
FORD           9199
BMW            7570
KIA            7432
TOYOTA         6288
VOLKSWAGEN     5004
JEEP           4480
HYUNDAI        4406
Name: count, dtype: int64


In [15]:
# Calculate the top 10 most common makes
top_models = df['Model'].value_counts().head(10)

print("Top 10 most common models:")
print(top_models)

Top 10 most common models:
Model
MODEL Y           35989
MODEL 3           30092
LEAF              13366
MODEL S            7735
BOLT EV            6821
MODEL X            5796
VOLT               4796
ID.4               3937
WRANGLER           3392
MUSTANG MACH-E     3322
Name: count, dtype: int64


# DATA CLEANING AND PREPROCESSING

## Handling missing values

We will be handling missing values by dropping rows with missing values for column County, City, Postal Code and Vehicle Location. For Electric Utility and 2020 Census Tract we will impute the missing values with value '0'

In [17]:
df = df.dropna(subset=['County', 'City', 'Postal Code', 'Vehicle Location'])

In [18]:
df['Electric Utility'].unique()

array(['CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)',
       'PUGET SOUND ENERGY INC',
       'PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)', 'PACIFICORP',
       'PUD NO 1 OF CHELAN COUNTY', 'MODERN ELECTRIC WATER COMPANY',
       'PUD NO 2 OF GRANT COUNTY', 'AVISTA CORP',
       'PUD NO 1 OF PEND OREILLE COUNTY',
       'BONNEVILLE POWER ADMINISTRATION||PUD NO 1 OF CLARK COUNTY - (WA)',
       'BONNEVILLE POWER ADMINISTRATION||PUD NO 1 OF COWLITZ COUNTY',
       'BONNEVILLE POWER ADMINISTRATION||PUD 1 OF SNOHOMISH COUNTY',
       'BONNEVILLE POWER ADMINISTRATION||PUGET SOUND ENERGY INC||PUD NO 1 OF JEFFERSON COUNTY',
       'NO KNOWN ELECTRIC UTILITY SERVICE',
       'BONNEVILLE POWER ADMINISTRATION||PUD NO 1 OF MASON COUNTY|PUD NO 1 OF JEFFERSON COUNTY',
       'BONNEVILLE POWER ADMINISTRATION||PACIFICORP||PUD NO 1 OF CLARK COUNTY - (WA)',
       'BONNEVILLE POWER ADMINISTRATION||PACIFICORP||BENTON RURAL ELECTRIC ASSN',
       'BONNEVILLE POWER ADMINISTRATION||PUD NO 1 OF CLALLAM 

In [19]:
df['2020 Census Tract'].unique()

array([5.30330078e+10, 5.30610519e+10, 5.30330068e+10, ...,
       2.40217508e+10, 2.40217526e+10, 2.10930014e+10])